pageserver/src/tenant/timeline.rs

```diff
@@ -2732,13 +2732,13 @@ impl Timeline {
         self.last_record_lsn.advance(new_lsn);
     }
 
-    fn freeze_inmem_layer(&self, write_lock_held: bool) {
+    async fn freeze_inmem_layer(&self, write_lock_held: bool) {
         // Freeze the current open in-memory layer. It will be written to disk on next
         // iteration.
         let _write_guard = if write_lock_held {
             None
         } else {
-            Some(self.write_lock.lock().unwrap())
+            Some(self.write_lock.lock().await)
         };
```

Comment:
During re-review of this patch, I couldn't spot the place where the caller that does `write_lock_held=true` actually acquired the write_lock.

@hlinnaka can you chime in?

Reply:
Huh, I can't see it either.

I tried to read the commit logs to understand how we got here. Looks like it's always been like that.

There is some discussion on 'write_lock' in `docs/rfcs/017-timeline-data-management.md`:
> - `write_lock: Mutex<()>` ensures that all timeline data being written into **in-memory layers** is done without races, one concurrent write at a time

I think the reason this hasn't caused any problems is that there is only one thread/task that can modify the in-memory layer: the walreceiver. And that task is also where check_checkpoint_distance() is called from. So in this case, we get away without holding the write_lock because there are no other tasks writing to the in-memory layer at the same time.

Sure looks wrong though.